submitit/core/core.py

```diff
@@ -289,6 +305,45 @@ def results(self) -> tp.List[R]:
             raise job_exception  # pylint: disable=raising-bad-type
         return [result]
 
+    async def async_results(self, poll_interval: tp.Union[int, float] = 1) -> tp.List[R]:
+        """ asyncio version of the results() method.
+
+        Waits asynchornously for ALL the results to be available by polling the self.done() method.
+
+        Parameters
+        ----------
+        poll_interval: int or float
+            how often to check if the result is available, in seconds
+        """
+        await self._async_wait(poll_interval)
+        # results are ready now
+        return self.results()
+
+    def results_as_compteled(
+        self, poll_interval: tp.Union[int, float] = 1
+    ) -> tp.Iterator["asyncio.Future[R]"]:
+        """awaits for all tasks results concurrently. Note that the order of results is not guaranteed to match the order
+            of the tasks anymore as the earliest task coming back might not be the first one you sent.
+
+            Returns
+            -------
+            an iterable of Awaitables that can be awaited on to get the earliest result available of the remaining tasks.
+
+            Parameters
+            ----------
+            poll_interval: int or float
+                how often to check if the result is available, in seconds
+
+            (see https://docs.python.org/3/library/asyncio-task.html#asyncio.as_completed)
+        """
+        if self._sub_jobs:
+            return asyncio.as_completed(
+                [async_cast(sub_job.async_result(poll_interval)) for sub_job in self._sub_jobs]
```

Comment:
Do we need async_cast ? It's my first time dealing with async code and mypy, but this seems a bit hacky.

https://mypy.readthedocs.io/en/stable/more_types.html#async-and-await

Reply:
yeah, not sure, I haven't tried without, I will. Future is actually not generically typed (hence the quoting) because it's part of python core :(